File: app/src/main/java/com/limelight/nvstream/av/video/VideoDecoderRenderer.java
package com.limelight.nvstream.av.video;

import java.nio.ByteBuffer;

public abstract class VideoDecoderRenderer {
    public abstract int setup(int format, int width, int height, int redrawRate);

    public abstract void start();

    public abstract void stop();

    // This is called once for each frame-start NALU. This means it will be called several times
    // for an IDR frame which contains several parameter sets and the I-frame data.
    public abstract int submitDecodeUnit(ByteBuffer decodeUnitData2, int decodeUnitLength, int decodeUnitType,
                                         int frameNumber, long receiveTimeMs);
    
    public abstract void cleanup();

    public abstract int getCapabilities();
}

File: app/src/main/java/com/limelight/nvstream/av/video/VideoDecoderRenderer.kt
package com.limelight.nvstream.av.video

abstract class VideoDecoderRenderer {
    abstract fun setup(format: Int, width: Int, height: Int, redrawRate: Int): Int

    abstract fun start()

    abstract fun stop()

    // This is called once for each frame-start NALU. This means it will be called several times
    // for an IDR frame which contains several parameter sets and the I-frame data.
    abstract fun submitDecodeUnit(
        decodeUnitData: ByteArray, decodeUnitLength: Int, decodeUnitType: Int,
        frameNumber: Int, frameType: Int, frameHostProcessingLatency: Char,
        receiveTimeUs: Long, enqueueTimeUs: Long, hostPresentationTimeUs: Long
    ): Int

    abstract fun cleanup()

    abstract fun getCapabilities(): Int

    abstract fun setHdrMode(enabled: Boolean, hdrMetadata: ByteArray?)

    // Called when the host resolution changes (e.g., screen rotation)
    open fun onResolutionChanged(width: Int, height: Int) {
        // Default implementation does nothing
        // Subclasses can override to handle resolution changes
    }
}
